src/v/cluster/topics_frontend.cc

```diff
@@ -794,4 +795,97 @@ ss::future<topic_result> topics_frontend::do_create_partition(
     }
 }
 
+ss::future<result<std::vector<partition_result>>>
+topics_frontend::cancel_moving_partition_replicas_node(
+  model::node_id node_id,
+  partition_move_direction dir,
+  model::timeout_clock::time_point timeout) {
+    using ret_t = result<std::vector<partition_result>>;
+    auto leader = _leaders.local().get_leader(model::controller_ntp);
+
+    // no leader available
+    if (!leader) {
+        co_return errc::no_leader_controller;
+    }
+    if (leader == _self) {
+        if (dir == partition_move_direction::from_node) {
+            co_return co_await do_cancel_moving_partition_replicas(
+              _topics.local().ntps_moving_from_node(node_id), timeout);
+        }
+
+        co_return co_await do_cancel_moving_partition_replicas(
+          _topics.local().ntps_moving_to_node(node_id), timeout);
+    }
+
+    co_return co_await _connections.local()
+      .with_node_client<controller_client_protocol>(
+        _self,
+        ss::this_shard_id(),
+        *leader,
+        timeout,
+        [timeout, node_id, dir](controller_client_protocol client) mutable {
+            return client
+              .cancel_node_partition_movements(
+                cancel_node_partition_movements_request{
+                  .node_id = node_id, .direction = dir},
+                rpc::client_opts(timeout))
+              .then(&rpc::get_ctx_data<cancel_partition_movements_reply>);
+        })
+      .then([](result<cancel_partition_movements_reply> r) {
+          return r.has_error() ? ret_t(r.error())
+                               : std::move(r.value().partition_results);
+      });
+}
+
+ss::future<result<std::vector<partition_result>>>
+topics_frontend::cancel_moving_all_partition_replicas(
+  model::timeout_clock::time_point timeout) {
+    using ret_t = result<std::vector<partition_result>>;
+    auto leader = _leaders.local().get_leader(model::controller_ntp);
+
+    // no leader available
+    if (!leader) {
+        co_return errc::no_leader_controller;
+    }
+    if (leader == _self) {
+        co_return co_await do_cancel_moving_partition_replicas(
+          _topics.local().all_updates_in_progress(), timeout);
+    }
+
+    co_return co_await _connections.local()
+      .with_node_client<controller_client_protocol>(
+        _self,
+        ss::this_shard_id(),
+        *leader,
+        timeout,
+        [timeout](controller_client_protocol client) mutable {
+            return client
+              .cancel_all_partition_movements(
```

Comment:
it seems that we have a potential for a redirect loop here if two nodes think of each other as leaders?

Reply:
this should never really happen, this would lead to split brain scenario, if the nodes are able to communicate and the request is actually exchanged between them they will reconcile the leadership state.